gh-106739: Add `rtype_cache` to `warnings.warn` message when leaked objects found

shailshouryya opened the pull request:
Closes Issue #106739: [Add rtype_cache to warnings.warn message when leaked objects found](https://github.com/python/cpython/issues/106739)


See message in commit a54a5fb84dba76aa42430addaa8914be2ac1da72 for the updated warning message.

<!-- gh-issue-number: gh-106739 -->
* Issue: gh-106739
<!-- /gh-issue-number -->

shailshouryya committed:
Add `rtype_cache` to `warnings.warn` message when leaked objects found

Adding the `rtype_cache` to the `warnings.warn` message improves the
previous, somewhat vague message from

```
/Users/username/cpython/Lib/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 6 leaked semaphore objects to clean up at shutdown
```

to

```
/Users/username/cpython/Lib/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 6 leaked semaphore objects to clean up at shutdown: {'/mp-yor5cvj8', '/mp-10jx8eqr', '/mp-eobsx9tt', '/mp-0lml23vl', '/mp-9dgtsa_m', '/mp-frntyv4s'}
```

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

pitrou commented on Misc/NEWS.d/next/Library/2023-07-20-06-00-35.gh-issue-106739.W1hygr.rst:
```diff
@@ -0,0 +1 @@
+Adds the `rtype_cache` `set` to the `warnings.warn` message (as an addition to the type of leaked objects and the number of leaked objects already included in the `warnings.warn` message) to make debugging leaked objects easier when the `resource_tracker` module (`Lib/multiprocessing/resource_tracker.py`) finds leaked objects. Including the names of the leaked objects will help more quickly ident ...
```
```suggestion
Add the ``rtype_cache`` to the warning message (as an addition to the type of leaked objects and the number of leaked objects already included in the message) to make debugging leaked objects easier when the multiprocessing resource tracker process finds leaked objects at shutdown. This helps more quickly identify what was leaked and/or why the leaked object was not properly cleaned up.
```

shailshouryya replied:
I updated the entry in commit 16ec55739400259a6b02b9a4553d1217d1fc56d1. Thanks for catching the incorrect formatting for `rtype_cache`! (I used markdown formatting and missed the double backticks)

pitrou commented on Lib/multiprocessing/resource_tracker.py:
```diff
@@ -222,8 +222,8 @@ def main(fd):
-                                  'leaked %s objects to clean up at shutdown' %
-                                  (len(rtype_cache), rtype))
+                                  'leaked %s objects to clean up at shutdown: %s' %
+                                  (len(rtype_cache), rtype, rtype_cache))
```
Not mandatory, but you could perhaps update this to use a f-string.

shailshouryya replied:
Sure thing, added in commit 819915e1d2ebe9d9069fd72a8675498bc1873a80 (and fixed the formatting in commit 1611dd4d8c3744e14bc72b7b7b53eef51b6bde96).

I kept the warning split out into multiple lines since keeping the entire message on 1 line resulted in 156 characters on a line, which seems a bit long. I couldn't find a formal formatting guide for multiline f-strings, so I currently have it formatted as
```
                    warnings.warn(
                        f'resource_tracker: There appear to be {len(rtype_cache)} '
                        f'leaked {rtype} objects to clean up at shutdown: {rtype_cache}'
                    )
```
but I wasn't sure if I should go with the following instead:
```
                    warnings.warn(f'resource_tracker: There appear to be {len(rtype_cache)} '
                                  f'leaked {rtype} objects to clean up at shutdown: {rtype_cache}')
```

Is there a preference for these kinds of scenarios?

pitrou replied:
I would go with the former in this case, to limit the indentation amount.

shailshouryya replied:
Okay, sounds good - in that case, this part should be good to go!

shailshouryya committed:
Update NEWS entry to be more concise and use proper formatting

shailshouryya committed:
Replace old style string formatting (% operator) with f-string formatting

shailshouryya committed:
Fix multiline f-string formatting

The previously used f-string format did not properly interpolate the
variables, and resulted in an output such as the following:

```
/Users/username/cpython/Lib/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 6 leaked {rtype} objects to clean up at shutdown: {rtype_cache}
```

Resource: https://stackoverflow.com/questions/49416042/how-to-write-an-f-string-on-multiple-lines-without-introducing-unintended-whites

pitrou reviewed (approved):
+1, thank you @shailshouryya
